google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -204,6 +205,10 @@ public B setHost(String host) {
      */
     public B setCredentials(Credentials credentials) {
       this.credentials = checkNotNull(credentials);
+      // set project id if available
+      if (credentials instanceof ServiceAccountCredentials) {
```

Comment:
In this case `serviceOptions = ServiceOptions.newBuilder().setCredentials(...).setProjectId(...).build()` and `serviceOptions = ServiceOptions.newBuilder().setProjectId(...).setCredentials(...).build()` might create `serviceOptions` with different `projectId`s. Probably set the `projectId` here only when `this.projectId == null`?

Reply:
Fixed and added tests.